src/celeritas/optical/action/detail/TrackInitAlgorithms.cu

```diff
@@ -6,25 +6,69 @@
 //---------------------------------------------------------------------------//
 #include "TrackInitAlgorithms.hh"
 
-#include <thrust/copy.h>
+// CUDA has included cub since CUDA 11, but ROCm does not include hipCUB by
+// default, so test for the availability of hipCUB and use thrust instead if
+// it's unavailable. And some further checks for newer cub/hipCUB functions.
+#if CELER_USE_HIP
+#    if CELERITAS_HAVE_HIPCUB
+#        define CELER_USE_THRUST 0
+#    else
+#        define CELER_USE_THRUST 1
+#    endif
+#endif
+#if CELERITAS_USE_CUDA
+#    include <cub/device/device_select.cuh>
+#    include <cub/version.cuh>
+#elif CELERITAS_USE_HIP && CELERITAS_HAVE_HIPCUB
+#    include <hipcub/device/device_select.hpp>
+#    include <hipcub/hipcub_version.hpp>
+#endif
+// DeviceTransform is unavailable in older versions of cub/hipcub, so fall back
+// to using thrust::transform instead
+#if CELERITAS_USE_CUDA && CUB_VERSION >= 200800
+#    define CELER_CUB_HAS_TRANSFORM 1
+#else
+#    define CELER_CUB_HAS_TRANSFORM 0
+#endif
+#if CELERITAS_USE_HIP && HIPCUB_VERSION >= 400100
```

Comment:
Some of these conditionals could be simplified a bit by using the fact that undefined macro symbols are set to zero in the preprocessor (or for clarity you could set them to zero yourself).